### PR TITLE
[10.x] Fix LazyCollection::only() to avoid closing nested generators

### PR DESCRIPTION
**Pull request by cerbero90**
This PR contains a fix for the `LazyCollection::only()` method.

When using `break` within *generators of generators*, we may end up closing nested generators. When this happens, we provoke a fatal error:

```
Cannot rewind a generator that was already run
```

Using `break` is safe when we are checking whether a generator is valid. If it's no longer valid, it makes sense to call break e.g.:

```php
// good use of break:

if (! $generator->valid()) {
    break;
}
```

Generators' design does not allow them to be run again once they are exhausted, that means that if we `break` the loop of a nested generator, the outer generator cannot iterate through it anymore and this provokes the previously mentioned fatal error.

The solution in this case is to continue the iteration without breaking it. When all the wanted keys are yielded, the condition `array_key_exists($key, $keys)` will check the key against an empty array. This will let us have an efficient iteration without the need of breaking it.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Not sure I'm comfortable just deleting code without understanding it better.

**Comment by cerbero90** [comment]
@taylorotwell I understand your point, generators are a complicated concept 👍

Please let me know what is not clear and I'll try to explain myself better

I hope this issue can be solved as the only() method is currently not working as expected

File: tests/Support/SupportLazyCollectionIsLazyTest.php [hunk 1]
```diff
@@ -782,11 +782,31 @@ public function testOnlyIsLazy()
...
+            ->map(fn (LazyCollection $nestedCollection) => $nestedCollection->only('foo', 'bar'))
+            ->toArray();
+
+        $expected = [
+            ['foo' => 1, 'bar' => 2],
+            ['foo' => 1, 'bar' => 2],
+        ];
+
+        $this->assertSame($expected, $actual);
+    }
```
Comment (JosephSilber) [inline tests/Support/SupportLazyCollectionIsLazyTest.php:808]: @cerbero90 This test seems to be passing even without the change above.

If I add back the `if...empty...break` block, this test still passes.

What am I missing?

**Comment by JosephSilber** [comment]
> Generators' design does not allow them to be run again once they are exhausted

Correct.

> that means that if we break the loop of a nested generator, the outer generator cannot iterate through it anymore and this provokes the previously mentioned fatal error.

Can you explain a little more? How is the outer generator affected by the inner `break`? When would it try to continue iterating over it after the `break`? Why would that be considered double enumeration?

**Pull request closed without merging**